Pattern library alerts on discussion pages

bjacobel opened the pull request:
## Description
 
[TNL-5164](https://openedx.atlassian.net/browse/TNL-5164)

Implement Pattern Library alerts in the discussion tab, inline discussions and the teams interface.

### Sandbox
- [x] https://bjacobel-discussion-alerts.sandbox.edx.org (updated Wed Sep  7 12:35:10 EDT 2016)

### Testing
- [ ] i18n
- [ ] RTL
- [x] safecommit violation code review process
- [x] Unit, integration, acceptance tests as appropriate
- [x] Analytics
- [x] Performance
- [x] Database migrations are backwards-compatible

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @andy-armstrong 
- [x] Code review: @dianakhuang 
- [x] Doc Review: @catong - no review required - FYI only
- [ ] UX review: @chris-mike 
- [ ] Accessibility review: @edx/edx-accessibility 
- [ ] Product review: @marcotuts 

### Post-review
- [ ] Rebase and squash commits

andy-armstrong commented on common/static/common/templates/discussion/new-post-alert.underscore:
```diff
@@ -0,0 +1,13 @@
+<div class="alert alert-error pattern-library-shim" role="alert" aria-labelledby="alert-title-error" tabindex="-1">
+    <span class="icon alert-icon fa fa-warning pattern-library-shim" aria-hidden="true"></span>
```
Do you need the pattern-library-shim class at every level? I thought it could just be included on the outermost div?

andy-armstrong commented on common/static/common/templates/discussion/new-post-alert.underscore:
```diff
@@ -0,0 +1,13 @@
+
+    <div class="alert-message pattern-library-shim">
+        <h3 class="sr-only alert-title pattern-library-shim" id="alert-title-error">
+            Error posting your message.
```
This should use ``gettext`` for i18n.

andy-armstrong commented on common/static/common/templates/discussion/new-post-alert.underscore:
```diff
@@ -0,0 +1,13 @@
+    <span class="icon alert-icon fa fa-warning pattern-library-shim" aria-hidden="true"></span>
+
+    <div class="alert-message pattern-library-shim">
+        <h3 class="sr-only alert-title pattern-library-shim" id="alert-title-error">
```
This id won't be unique if there are two discussion components shown in the same unit. I don't know that this is very common, but can it use a unique id?

andy-armstrong commented on lms/static/sass/discussion/views/_create-edit-post.scss:
```diff
@@ -11,6 +11,11 @@
   max-width: 1180px;
 
+  .post-errors {
+    padding: 0 !important;
```
Are the ``!important``s necessary? I think the intention is to never use them except in the shame file. Can you make them more specific?

bjacobel replied:
These are the styles we're overriding:
![](
![image](https://cloud.githubusercontent.com/assets/407639/18138043/ddfb770c-6f78-11e6-9290-7d05e18b5714.png)) (from [here](https://github.com/edx/edx-platform/blob/bjacobel/discussion-alerts/lms/static/sass/course/courseware/_courseware.scss#L479)) so as you can see we'd have to get quite a bit more specific (like adding an additional three selectors or something) to override those styles naturally. I don't like `!important`s very much either, but I think we need them 😓  I could move them to the shame file, though.

andy-armstrong replied:
Uggh, those rules directly on elements are terrible! I think what you have makes sense then. If you want to move them to shame that would be great, or you could just leave them here.

marcotuts replied:
Man.... I hate .vert-mod and all xmodule styles in the platform. :/

andy-armstrong commented on lms/static/sass/shared/_alerts_pattern_library_shim.scss:
```diff
@@ -19,38 +19,49 @@
-$alert-border: 0.0625rem solid $alert-border-grey;
+// Overrides for other things available in the UXPL but not here
+// This is super gross. Don't do this in other places if you can avoid it.
+$palette-brand-accent: #2991c3;
```
This is problematic for anyone who is going to theme these pages. I wonder if there's a cleaner way to be able to consume these colors.

bjacobel replied:
We could convert courseware to the pattern library and just kill this file.

andy-armstrong replied:
As we discussed at stand up, I understand now that this is just shimming in the same code from the Pattern Library. 👍 for now, and as you say we should convert courseware to the pattern library in the medium term.

andy-armstrong commented on lms/static/sass/shared/_alerts_pattern_library_shim.scss:
```diff
@@ -19,38 +19,49 @@
     .alert-icon {
-        color: $alert-icon-color;
-        background-color: $alert-color;
+        color: $white;
```
It would be better to use a variable so that it can be overridden by a theme.

andy-armstrong commented on lms/static/sass/shared/_alerts_pattern_library_shim.scss:
```diff
@@ -60,31 +71,62 @@ $alert-border: 0.0625rem solid $alert-border-grey;
-        border-top-left-radius: 0;
-        border-top-right-radius: 0;
-        padding: 1.25rem;
+        background-color: $white;
```
Again, use a variable here.

bjacobel replied:
[The UXPL is using $white here](https://github.com/edx/ux-pattern-library/blob/master/pattern-library/sass/patterns/_alerts.scss#L44), isn't it better to keep differences between that file and this one to a minimum?

andy-armstrong replied:
Agreed, let's just keep the shim the same. I'll make a ticket to clean up how we will theme the pattern library itself.

andy-armstrong replied:
Here's the ticket: https://openedx.atlassian.net/browse/FEDX-236

bjacobel commented:
Fixed tests and addressed comments.

andy-armstrong commented:
👍

dianakhuang commented:
👍  from me as well

bjacobel commented:
So the pattern we've implemented here is (theoretically) already ux and a11y reviewed, and there's no need for a doc change... is this good to go, do people think?

andy-armstrong commented:
I think you just need a thumb from @clrux as the reported issue is that we didn't give correct focus to the alerts.

clrux commented:
Are we sending focus now?

bjacobel commented:
We're currently not in production, I believe I fixed it in this PR but didn't write any tests for it.

(FYI the stager is updating right now so things may be wonky until like 4:15)

catong commented:
No need for doc review on this one. Thanks!

bjacobel commented:
@clrux bumping this, if you wouldn't mind finishing the review - I think the sandbox wasn't ready last time when you started, but it should be up to date now.

clrux commented on common/static/common/templates/discussion/new-post.underscore:
```diff
@@ -1,5 +1,5 @@
 <form class="forum-new-post-form">
-    <ul class="post-errors" style="display: none"></ul>
+    <ul class="post-errors" style="display: none" tabindex="-1"></ul>
```
We don't need the unordered list to be focusable, only the specific message we're sending focus to when it appears.

clrux commented:
@bjacobel Will we ever have more than one alert at a time? If not, we may want to use something other than an unordered list for the wrapper. A plain `div` will suffice. Everything else here - the role, the aria-labelledby, etc. is good.

I was only able to check the inline discussion; I can't get to the full-on discussion tab, and I don't see a Teams tab here either. I'm assuming the implementation is the same though.

bjacobel commented:
@clrux The error displaying logic is set up to display all the errors the backend sends if a request fails... in my testing I wasn't able to get it to encounter multiple errors at once, but the backend does communicate errors back to the frontend in an array, so it's theoretically possible that there might be more than one (and seems designed for that case). I think it's best to leave the `ul/li` for now because we don't know for sure it'll _never_ be more than one entry in that list.

I'll shift the focus to the individual `li` rather than the containing `ul`, though. Thanks for the review!

clrux commented:
I'd send focus to the `role="alert"`, so the `div` in this case, yes?

bjacobel commented:
@clrux Sounds good.

bjacobel commented:
Updated to send focus properly, building another sandbox now.

FYI, this also fixes TNL-5436 - noticed it happening on this branch and fixed it before I saw the separate bug filed.

bjacobel commented:
IDK what's up with Coveralls, I didn't change test coverage at all here.
